app/src/main/java/de/westnordost/streetcomplete/data/osm/mapdata/MapDataCache.kt

```diff
@@ -239,13 +248,14 @@ class MapDataCache(
     /** Gets the nodes with the given [ids] from cache. If any of the nodes are not cached, [fetch]
      *  is called for the missing nodes. */
     fun getNodes(ids: Collection<Long>, fetch: (Collection<Long>) -> List<Node>): List<Node> = synchronized(this) {
-        val cachedNodes = spatialCache.getAll(ids)
+        val cachedNodes = ids.mapNotNull { spatialCache.get(it) ?: nodeCache[it] }
```

Comment:
This should be slower than 
```
spatialCache.getAll(ids) + ids.mapNotNull { nodeCache[it] }
```
due to the `synchronized` in `SpatialCache.get` but will return duplicate nodes if `nodeCache` is not guaranteed to be disjunct from `SpatialCache`.

Reply:
This should work, as on `update` all nodes contained in `spatialCache` are removed from `nodeCache`

Reply:
On the other hand, `spatialCache.get` doesn't seem to be significantly slowed down by `synchronized`, and this way looping over ids twice can be avoided.

Though I think it's not really relevant here. Even for 4k nodes it would take 5 ms on my ancient phone, and in my testing it looks like `getNodes` is usually called for something like 3-10 node ids, occasionally also 50 or 60.

I had to use microseconds to actually measure the time this usually takes, but didn't find any noticeable difference between `ids.mapNotNull { spatialCache.get(it) ?: nodeCache[it] }` and `spatialCache.getAll(ids) + ids.mapNotNull { nodeCache[it] }`